### PR TITLE
[Feat] Initialize UAPI

### PR DESCRIPTION
**Pull request by Bobholamovic**
### PR types
New features

### PR changes
APIs

### CODE REVIEW

File: uapi/base/config.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        raise NotImplementedError
+
+    def load(self, config_file_path):
+        raise NotImplementedError
+
+    def dump(self, config_file_path):
+        raise NotImplementedError
+
+    @classmethod
+    def build_from_file(cls, config_file_path, *args, **kwargs):
```
Comment (WenmuZhou) [inline uapi/base/config.py]: cls和上面的方法统一用self吧

File: uapi/base/utils/download.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (WenmuZhou) [inline uapi/base/utils/download.py]: 这个文件没用到可以先删掉

File: uapi/model_zoo.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from .base.register import register_arch_info, register_model_info
+from .seg_task import SegModel, SegRunner
+
+# XXX: Hard-code relative path of repo root dir
+REPO_ROOT_PATH = osp.abspath(osp.join(osp.dirname(__file__), '..'))
+register_model_info({
+    'model_name': 'SegModel',
+    'model_cls': SegModel,
+    'runner_cls': SegRunner,
+    'repo': 'PaddleSeg',
```
Comment (WenmuZhou) [inline uapi/model_zoo.py]: repo属性，后面好像没用到

File: uapi/base/config.py (hunk 1 above)
Comment (Bobholamovic) [reply uapi/base/config.py]: 这个主要是为了 `Config`.build_from_file() 的用法，返回一个新对象，如果是实例方法的话得写成：`Config().build_from_file()`，感觉是不是前一种更简洁一些？

File: uapi/base/utils/download.py (hunk 2 above)
Comment (Bobholamovic) [reply uapi/base/utils/download.py]: done.

File: uapi/model_zoo.py (hunk 3 above)
Comment (Bobholamovic) [reply uapi/model_zoo.py]: 已删除。

File: uapi/base/config.py (hunk 1 above)
Comment (WenmuZhou) [reply uapi/base/config.py]: Config.build_from_file() 这种用法是不是@staticmethod 就可以

File: uapi/base/config.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update_weight_decay(self, weight_decay):
+        """Update configurations of weight decay."""
+        raise NotImplementedError
+
+    @classmethod
+    def build_from_file(cls, config_file_path, *args, **kwargs):
```
Comment (juncaipeng) [inline uapi/base/config.py]: Config需要考虑直接暴露给用户的用法，用户可以设置name和config_path，然后更新信息，将Config传给PaddleModel。

File: uapi/base/model.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+from .register import (get_registered_model_info, build_runner_from_model_info,
+                       build_model_from_model_info)
+from .utils.misc import CachedProperty as cached_property
+from .utils.path import create_yaml_config_file
+
+
+class PaddleModel(object):
+    # We constrain function params here
+    def __new__(cls, model_name):
```
Comment (juncaipeng) [inline uapi/base/model.py]: new需要加入config输入参数，model_name和config必须设置一个，然后再构建model。

File: uapi/seg/register.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+# XXX: Hard-code relative path of repo root dir
+REPO_ROOT_PATH = osp.abspath(osp.join(osp.dirname(__file__), '..', '..'))
+register_suite_info({
+    'suite_name': 'Seg',
+    'model': SegModel,
+    'runner': SegRunner,
+    'runner_root_path': REPO_ROOT_PATH,
+    'supported_api_list':
+    ['train', 'predict', 'export', 'infer', 'compression']
```
Comment (juncaipeng) [inline uapi/seg/register.py]: 需要对于每一个model注册支持的api，可能有的模型无法进行导出和压缩

File: uapi/base/model.py (hunk 5 above)
Comment (Bobholamovic) [reply uapi/base/model.py]: done

File: uapi/seg/register.py (hunk 6 above)
Comment (Bobholamovic) [reply uapi/seg/register.py]: done

File: uapi/base/config.py (hunk 4 above)
Comment (Bobholamovic) [reply uapi/base/config.py]: done

File: uapi/base/register.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def build_config_from_model_info(model_info, config_file_path=None):
+    suite_name = model_info['suite']
+    # `suite_name` being the primary key of suite info
+    suite_info = get_registered_suite_info(suite_name)
+    config_cls = suite_info['config']
+    if config_file_path is None:
+        config_file_path = model_info['config_path']
+    config_obj = config_cls.build_from_file(config_file_path)
```
Comment (juncaipeng) [inline uapi/base/register.py]: config采用常规方式进行构建

File: uapi/base/config.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+from .register import get_registered_model_info, build_config_from_model_info
+
+__all__ = ['Config', 'BaseConfig']
+
+
+class Config(object):
+    # We constrain function params here
+    def __new__(cls, model_name, config_file_path=None):
+        model_info = get_registered_model_info(model_name)
+        return build_config_from_model_info(
```
Comment (juncaipeng) [inline uapi/base/config.py]: 此处考虑config_file_path和model_name的覆盖逻辑吧，不要在build_config_from_model_info里面了

File: uapi/seg/model.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    def train(self,
+              dataset=None,
+              batch_size=None,
+              learning_rate=None,
+              epochs_iters=None,
+              device=None,
+              resume_path=None,
+              dy2st=None,
+              amp=None,
+              save_dir=None):
```
Comment (juncaipeng) [inline uapi/seg/model.py:37]: 改下所有函数的默认参数，device='gpu', dy2st=False

Comment (Bobholamovic) [reply uapi/seg/model.py:37]: done

File: uapi/base/config.py (hunk 8 above)
Comment (Bobholamovic) [reply uapi/base/config.py]: done

File: uapi/base/register.py (hunk 7 above)
Comment (Bobholamovic) [reply uapi/base/register.py]: done

**Review by WenmuZhou — DISMISSED** [review]

File: uapi/base/config.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    Args:
+        model_name (str): A registered model name.
+        config_file_path (str|None): Path of a configuration file.
+        cfg (BaseConfig|None): `BaseConfig` object to initialize from.
+    """
+
+    def __init__(self, model_name, config_file_path=None, cfg=None):
+        super().__init__(cfg=cfg)
+        self.model_name = model_name
+        model_info = get_registered_model_info(self.model_name)
```
Comment (juncaipeng) [inline uapi/base/config.py]: 这行需要放到下面的if里面

File: uapi/base/model.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            epochs_iters (int|None): Total iterations or epochs of model training. If None, use a 
+                pre-defined default value of epochs/iterations.
+            device (str|None): A string that describes the device(s) to use, e.g., 'cpu', 'xpu:0', 'gpu:1,2'. 
+            save_dir (str|None): Directory to store model snapshots. The exported model will be saved in the 
+                `infer` subdirectory of `save_dir`. If None, use a default setting. 
+        """
+        raise NotImplementedError
+
+    @cached_property
+    def config_file_path(self):
```
Comment (juncaipeng) [inline uapi/base/model.py]: _config_file_path

File: uapi/base/config.py (hunk 10 above)
Comment (Bobholamovic) [reply uapi/base/config.py]: done

File: uapi/base/model.py (hunk 11 above)
Comment (Bobholamovic) [reply uapi/base/model.py]: done

**Comment by Bobholamovic** [comment]
PaddleX模型介绍配图：

DINO
FCOS
Fast-SCNN
PP-HumanSeg
MaskFormer
Faster R-CNN
Mask R-CNN
YOLOv3
